src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -303,21 +277,133 @@ private void AddExtensionAssembliesFromSource(IEnumerable<string> sources)
+                .SelectMany(path => ExpandTestAdapterPaths(path, fileHelper, strategy))
+                .Union(defaultExtensionPaths);
+        }
+
+        var extensionsFolder = Path.Combine(Path.GetDirectoryName(typeof(TestPlatform).GetTypeInfo().Assembly.GetAssemblyLocation()), "Extensions");
         if (fileHelper.DirectoryExists(extensionsFolder))
         {
-            var defaultExtensionPaths = fileHelper.EnumerateFiles(
-                extensionsFolder,
-                SearchOption.TopDirectoryOnly,
-                ".dll",
-                ".exe");
+            // Load default runtime providers
+            if ((strategy & TestAdapterLoadingStrategy.DefaultRuntimeProviders) == TestAdapterLoadingStrategy.DefaultRuntimeProviders)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, TestPlatformConstants.RunTimeEndsWithPattern)
+                    .Union(defaultExtensionPaths);
+            }
+
+            // Default extension loader
+            if (strategy == TestAdapterLoadingStrategy.Default
+                || (strategy & TestAdapterLoadingStrategy.ExtensionsDirectory) == TestAdapterLoadingStrategy.ExtensionsDirectory)
+            {
+                defaultExtensionPaths = fileHelper
+                    .EnumerateFiles(extensionsFolder, SearchOption.TopDirectoryOnly, ".dll", ".exe")
+                    .Union(defaultExtensionPaths);
+            }
+        }
+
+        TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths.Distinct();
+    }
+
+    private static SearchOption GetSearchOption(TestAdapterLoadingStrategy strategy, SearchOption defaultStrategyOption) {
+        if (strategy == TestAdapterLoadingStrategy.Default) {
+            return defaultStrategyOption;
+        }
+
+        var searchOption = SearchOption.TopDirectoryOnly;
+        if ((strategy & TestAdapterLoadingStrategy.Recursive) == TestAdapterLoadingStrategy.Recursive)
+        {
+            searchOption = SearchOption.AllDirectories;
+        }
+
+        return searchOption;     
+    }
 
-            TestPluginCache.Instance.DefaultExtensionPaths = defaultExtensionPaths;
+    private static IEnumerable<string> ExpandTestAdapterPaths(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
+    {
+        var adapterPath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(path));
+
+        // Default behavior is to only accept directories!
+        if (strategy == TestAdapterLoadingStrategy.Default)
+        {
+            return ExpandAdaptersWithDefaultStrategy(adapterPath, fileHelper);
+        }
+
+        var adapters = ExpandAdaptersWithExplicitStrategy(adapterPath, fileHelper, strategy);
+
+        return adapters.Distinct();
+    }
+
+    private static IEnumerable<string> ExpandAdaptersWithExplicitStrategy(string path, IFileHelper fileHelper, TestAdapterLoadingStrategy strategy)
```

Comment:
Feel free to ignore but I would rename `path` into `fileOrDirectoryPath` (or `fileOrFolderPath`).